**DATA MANIPULATION**

In this notebook, I perform exercise of data manipulation for `telco_customer_churn.csv` dataset. Data Manipulation aims to adjust the data to become  more understandable for generating analysis

In [4]:
import pandas as pd
import numpy as np

After import the libraries needed, i need to read the data set using pandas. The source of dataset comes from kaggle https://www.kaggle.com/datasets/blastchar/telco-customer-churn

In [5]:
df = pd.read_csv('telco_customer_churn.csv')

In [6]:
df

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


**1. Remove rows with `tenure` = 0 , how many rows are currently available in the data ?**

In [9]:
df = df[df['tenure']!=0].reset_index(drop=True)
df['TotalCharges'] = df['TotalCharges'].astype(float)

In [10]:
df.shape

(7032, 21)

The rows are currently available are 7032 rows

**2. What is the average of `tenure` for customer who have no partner and no dependents ?**

answer : 23.303354

**3. What is the maximum `MonthlyCharges` recorded for customer who have partner and no dependents ?**

answer: 11860

**4. Count the customers who have dependents but no partner !**

answer : 359

In [13]:
#Code for answering number 2-4

In [12]:
df.groupby(['Partner', 'Dependents']).agg({'tenure':['mean', 'median'],
                                           'MonthlyCharges':['max', 'min'],
                                           'customerID':'count'})

tenure        MonthlyCharges        customerID
                         mean median            max    min      count
Partner Dependents                                                   
No      No          23.303354   16.0         118.65  18.25       3280
        Yes         23.986072   18.0         116.15  19.10        359
Yes     No          42.709014   47.0         118.60  18.40       1653
        Yes         41.578161   45.0         118.75  18.70       1740

**5. What is the average of `TotalCharges` from non-senior citizen customers (Senior citizen = 0) who have both `PhoneServices` and `Internet Sercive Fiber Optic`**

answer : 3226.492737

In [16]:
df.pivot_table(index = ['SeniorCitizen', 'PhoneService'],
              columns = ['InternetService'],
              values = ['TotalCharges'], aggfunc = 'mean')

TotalCharges                         
InternetService                     DSL  Fiber optic          No
SeniorCitizen PhoneService                                      
0             No            1508.513021          NaN         NaN
              Yes           2344.751771  3226.492737  655.759775
1             No            1455.687019          NaN         NaN
              Yes           2542.344194  3147.553430  932.299038